### DISCUSSION TITLE
Deserializing maps with enum keys

### DISCUSSION THREAD

**Discussion by onavratil-monetplus** [nlohmann/json#2410] [Ideas]
The library provides cool features to work with enums/enum classes. Lets have
```cpp
enum class MyEnum{
        INVALID,
        ONE,
        TWO
    };
    NLOHMANN_JSON_SERIALIZE_ENUM(MyEnum, {
        {MyEnum::INVALID, nullptr},
        {MyEnum::ONE, "one"},
        {MyEnum::TWO, "two"}
    })
```
Now I want to have a map whose keys are of the `enum class` type. I can do something like....
```cpp
auto obj = "{\"one\": 123}"_json;
auto myMap = obj.get<std::map<std::string, int>>();
for(const auto &[k, v] : myMap){
    MyEnum key = nlohmann::json(k).get<MyEnum>();
}
```
However, if i try a easier way
```cpp
auto myMapEasier =  obj.get<std::map<MyEnum, int>>();
```
...i get a strange error
```[json.exception.type_error.302] type must be array, but is object```
I have read through some issues, but mostly, they are concerned about serialization of such a map, not the opposite way.

Is this a desired behavior? Or is there some easier workaround I am missing?

**Comment by gregmarr** [comment]
I believe that this is because the type `std::map<NonString, something>` is serialized by the library as if it were `std::vector<std::pair<NonString, something>>`.

**Comment by onavratil-monetplus** [comment]
Thank you for the answer, I understand it much better now. Just to be a little more verbose...

The problem lies in the fact that my `enum class` indeed is not directly convertible to string. It can be converted via json thanks to the `NLOHMANN_JSON_SERIALIZE_ENUM` macro, but this is not a direct enum to string cast. I did this in
`nlohmann::json(k).get<MyEnum>();`  (first create json and then convert it to string).

The library indeed makes the difference based on `std::is_convertible<KeyType,std::string>::value` which does not take any detours. For now, the important fact is that the "roundtrip holds", i.e. a `std::map<MyEnum, int>`, when serialized by `nlohmann::json`, will be represented as an array of pairs (2-element arrays), which is exactly the format that is expected and works for deserialization.

The only way out of this would be to have an (implicit or explicit) conversion of `MyEnum` to `std::string`, however, state-of-art C++, as far as i know, offers no way to overload cast operators for `enum class`
